File: frontend/src/lib/portal-api.ts
import 'server-only'
import { auth } from '@clerk/nextjs/server'
import { headers } from 'next/headers'
import { resolveHost } from '@/lib/host'
import { appDomain } from '@/lib/domain'
import type {
  PortalAppointmentDetail,
  PortalAppointmentsResponse,
  PortalAvailabilityResponse,
  PortalBookingServicesResponse,
  PortalChallengeResponse,
  PortalContextResponse,
  PortalFinanceResponse,
  PortalMessagesResponse,
  PortalPetDetail,
  PortalPreferencesResponse,
  PortalPetSummary,
  PortalReceiptResponse,
  PortalStatementResponse,
  PortalTaxiOffer,
  PortalTaxiRide,
  PortalTenantResponse,
  PortalTimelineResponse,
  UpdateOwnPetInput,
  UpdatePortalPreferenceInput,
} from '@petshop/shared-types'

/**
 * Cliente do Portal, para uso no servidor.
 *
 * Separado de `lib/api.ts` por uma diferença que não é de estilo: **o Portal precisa
 * dizer de que petshop se fala**, e o Admin não. A sessão da equipe carrega a
 * Organization do Clerk, e dela o gateway resolve o tenant; o tutor não é membro de
 * organização nenhuma, então quem sabe o petshop é o host que este processo acabou de
 * servir — `petshopdojoao.{APP_DOMAIN}`. O slug sai daqui e vai no header.
 *
 * O token continua morando no servidor, como no Admin: o browser nunca fala com o
 * gateway.
 */

const baseUrl =
  process.env.API_URL ?? process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:3000'

const TENANT_SLUG_HEADER = 'x-petshop-tenant-slug'
const REQUEST_TIMEOUT_MS = 15_000

/**
 * O petshop deste endereço.
 *
 * Em desenvolvimento não há subdomínio (`localhost:3002` é sempre Admin), então o slug
 * vem de `PORTAL_DEV_SLUG` — é o que permite abrir o Portal na máquina de quem
 * desenvolve sem mexer no `/etc/hosts` a cada tenant novo.
 */
export async function portalSlug(): Promise<string | null> {
  const host = (await headers()).get('host') ?? ''
  const resolved = resolveHost(host, appDomain())
  if (resolved.slug) return resolved.slug
  return process.env.PORTAL_DEV_SLUG ?? null
}

export class PortalError extends Error {
  constructor(
    readonly status: number,
    readonly code: string,
    message: string,
    /**
     * O resto do `problem+json`, sem a moldura.
     *
     * O BFF manda contexto junto de vários erros — os horários próximos de um conflito,
     * os alertas clínicos a reconhecer, o valor da taxa, as janelas alternativas do
     * leva-e-traz. Sem este campo, a tela recebia a frase e jogava fora o que a tornava
     * acionável.
     */
    readonly extra: Record<string, unknown> = {},
  ) {
    super(message)
  }

  /** A lista de horários que o 409 de van cheia oferece (AC-04 de MOD-PORTAL-07). */
  get alternativeStartsAt(): string[] | undefined {
    const value = this.extra.alternativeStartsAt
    return Array.isArray(value) ? (value as string[]) : undefined
  }
}

interface RequestOptions {
  method?: 'GET' | 'POST' | 'PATCH'
  path: string
  body?: unknown
  /** Rota pública do Portal: a identidade visual, antes de haver sessão. */
  anonymous?: boolean
}

async function request<T>(options: RequestOptions): Promise<T> {
  const slug = await portalSlug()
  if (!slug) {
    throw new PortalError(404, 'ERR_PORTAL_001', 'Estabelecimento não encontrado')
  }

  const requestHeaders: Record<string, string> = {
    'content-type': 'application/json',
    [TENANT_SLUG_HEADER]: slug,
  }

  if (!options.anonymous) {
    const session = await auth()
    // O template nomeado publica o claim que o gateway compara; sem ele, cai no token
    // de sessão padrão. Mesma regra do Admin (ver `docs/setup-clerk.md`).
    const token =
      (await session.getToken({ template: 'petshop' }).catch(() => null)) ??
      (await session.getToken())
    if (token) requestHeaders.authorization = `Bearer ${token}`
  }

  /**
   * Teto por requisição, pelo mesmo motivo de `lib/api.ts`: um `await` que nunca volta
   * dentro de um Server Component deixa a página **em branco, sem erro nenhum** — nem
   * overlay, nem linha de log —, e não há o que investigar depois.
   */
  const controller = new AbortController()
  const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS)

  try {
    const response = await fetch(`${baseUrl}${options.path}`, {
      method: options.method ?? 'GET',
      headers: requestHeaders,
      ...(options.body !== undefined ? { body: JSON.stringify(options.body) } : {}),
      signal: controller.signal,
      cache: 'no-store',
    })

    const payload = (await response.json().catch(() => null)) as
      | ({ code?: string; detail?: string } & Record<string, unknown>)
      | null

    if (!response.ok) {
      // A moldura do problem+json fica de fora; o que interessa é o contexto do erro.
      const moldura = new Set(['code', 'detail', 'title', 'status', 'type'])
      const extra = Object.fromEntries(
        Object.entries(payload ?? {}).filter(([chave]) => !moldura.has(chave)),
      )

      throw new PortalError(
        response.status,
        payload?.code ?? 'ERR_PORTAL_010',
        payload?.detail ?? 'Não foi possível concluir agora. Tente novamente.',
        extra,
      )
    }

    return payload as T
  } catch (error) {
    if (error instanceof PortalError) throw error
    if (controller.signal.aborted) {
      throw new PortalError(504, 'ERR_PORTAL_010', 'O sistema demorou para responder.')
    }
    throw new PortalError(502, 'ERR_PORTAL_010', 'Não foi possível falar com o sistema.')
  } finally {
    clearTimeout(timeout)
  }
}

export function readPortalTenant(): Promise<PortalTenantResponse> {
  return request({ path: '/portal/v1/tenant', anonymous: true })
}

export function readPortalContext(): Promise<PortalContextResponse> {
  return request({ path: '/portal/v1/me' })
}

export function requestAccessCode(body: {
  identifier: string
  website?: string
}): Promise<PortalChallengeResponse> {
  return request({ method: 'POST', path: '/portal/v1/access/challenge', body })
}

export function verifyAccessCode(body: {
  challengeId: string
  code: string
}): Promise<PortalContextResponse> {
  return request({ method: 'POST', path: '/portal/v1/access/verify', body })
}

export function readOwnPets(): Promise<{ pets: PortalPetSummary[] }> {
  return request({ path: '/portal/v1/pets' })
}

export function readOwnPet(petId: string): Promise<PortalPetDetail> {
  return request({ path: `/portal/v1/pets/${petId}` })
}

export function updateOwnPet(
  petId: string,
  body: UpdateOwnPetInput,
): Promise<PortalPetDetail> {
  return request({ method: 'PATCH', path: `/portal/v1/pets/${petId}`, body })
}

/**
 * O histórico, paginado por cursor.
 *
 * O cursor é a hora do último atendimento entregue, e vai na URL já codificado: um
 * `+` de fuso horário não sobrevive à query string sem isso, e a página seguinte
 * voltaria vazia sem erro nenhum.
 */
export function readOwnPetTimeline(
  petId: string,
  options: { cursor?: string; limit?: number } = {},
): Promise<PortalTimelineResponse> {
  const query = new URLSearchParams()
  if (options.cursor) query.set('cursor', options.cursor)
  if (options.limit) query.set('limit', String(options.limit))
  const suffix = query.size > 0 ? `?${query.toString()}` : ''

  return request({ path: `/portal/v1/pets/${petId}/timeline${suffix}` })
}

// ─── MOD-PORTAL-05 — Agendamento Online ──────────────────────────────────────

export function readBookableServices(petId: string): Promise<PortalBookingServicesResponse> {
  return request({ path: `/portal/v1/booking/services?petId=${petId}` })
}

/**
 * Os horários de um dia.
 *
 * `date` é dia civil no fuso do petshop, e não um instante: quem escolhe escolhe um
 * dia, e converter para UTC aqui jogaria a madrugada para o dia anterior.
 */
export function readAvailability(query: {
  petId: string
  serviceIds: string[]
  date: string
}): Promise<PortalAvailabilityResponse> {
  const search = new URLSearchParams({
    petId: query.petId,
    serviceIds: query.serviceIds.join(','),
    date: query.date,
  })
  return request({ path: `/portal/v1/booking/availability?${search.toString()}` })
}

export interface CreatedBookingResponse {
  id: string
  status: string
  startsAt: string
  endsAt: string
  petName: string
  professionalName: string
  services: string[]
  totalCents: number
  awaitingApproval: boolean
  duplicate: boolean
  taxi: PortalTaxiRide[]
  taxiWarning: string | null
}

export function createBooking(body: {
  petId: string
  serviceIds: string[]
  startsAt: string
  professionalId: string
  acknowledgedAlerts?: boolean
  taxi?: { pickup: boolean; dropoff: boolean }
}): Promise<CreatedBookingResponse> {
  return request({ method: 'POST', path: '/portal/v1/booking', body })
}

// ─── MOD-PORTAL-07 — Taxi Dog no Agendamento ─────────────────────────────────

/**
 * A oferta de leva-e-traz para este tutor.
 *
 * Não depende do pet nem do horário escolhido: o preço é do **CEP** do endereço
 * primário, e é por isso que a pergunta cabe uma vez só por tela em vez de a cada
 * mudança de serviço.
 */
export function readTaxiOffer(): Promise<PortalTaxiOffer> {
  return request({ path: '/portal/v1/booking/taxi' })
}

// ─── MOD-PORTAL-06 — Meus Agendamentos ───────────────────────────────────────

export function readOwnAppointments(
  options: { cursor?: string; limit?: number } = {},
): Promise<PortalAppointmentsResponse> {
  const query = new URLSearchParams()
  if (options.cursor) query.set('cursor', options.cursor)
  if (options.limit) query.set('limit', String(options.limit))
  const suffix = query.size > 0 ? `?${query.toString()}` : ''

  return request({ path: `/portal/v1/appointments${suffix}` })
}

export function readOwnAppointment(id: string): Promise<PortalAppointmentDetail> {
  return request({ path: `/portal/v1/appointments/${id}` })
}

export function cancelOwnAppointment(
  id: string,
  body: { acknowledgeFee: boolean },
): Promise<PortalAppointmentDetail> {
  return request({ method: 'POST', path: `/portal/v1/appointments/${id}/cancel`, body })
}

export function rescheduleOwnAppointment(
  id: string,
  body: { startsAt: string; professionalId: string },
): Promise<PortalAppointmentDetail> {
  return request({ method: 'POST', path: `/portal/v1/appointments/${id}/reschedule`, body })
}

// ─── MOD-PORTAL-08 — Extrato e Recibos ───────────────────────────────────────

export function readOwnFinance(): Promise<PortalFinanceResponse> {
  return request({ path: '/portal/v1/finance' })
}

/**
 * O extrato, paginado por página.
 *
 * Por página e não por cursor, ao contrário do histórico do pet: os lançamentos são
 * ordenados por data do fato gerador, que **repete** quando três serviços do mesmo dia
 * entram juntos. Um cursor por data pularia linhas ou as repetiria.
 */
export function readOwnStatement(
  options: { page?: number; limit?: number } = {},
): Promise<PortalStatementResponse> {
  const query = new URLSearchParams()
  if (options.page) query.set('page', String(options.page))
  if (options.limit) query.set('limit', String(options.limit))
  const suffix = query.size > 0 ? `?${query.toString()}` : ''

  return request({ path: `/portal/v1/finance/statement${suffix}` })
}

export function readOwnReceipt(paymentId: string): Promise<PortalReceiptResponse> {
  return request({ path: `/portal/v1/finance/receipts/${paymentId}` })
}

// ─── MOD-PORTAL-10 — Central de Comunicação ──────────────────────────────────

/**
 * O histórico de mensagens, paginado por página como o extrato.
 *
 * Pelo mesmo motivo de lá: uma campanha enfileira centenas de mensagens no mesmo
 * instante, e um cursor por data pularia ou repetiria linhas dentro do lote.
 */
export function readOwnMessages(
  options: { page?: number; limit?: number } = {},
): Promise<PortalMessagesResponse> {
  const query = new URLSearchParams()
  if (options.page) query.set('page', String(options.page))
  if (options.limit) query.set('limit', String(options.limit))
  const suffix = query.size > 0 ? `?${query.toString()}` : ''

  return request({ path: `/portal/v1/messages${suffix}` })
}

export function readOwnPreferences(): Promise<PortalPreferencesResponse> {
  return request({ path: '/portal/v1/preferences' })
}

/** Um canal por chamada: cada clique é uma transição, e a trilha é append-only. */
export function updateOwnPreference(
  body: UpdatePortalPreferenceInput,
): Promise<PortalPreferencesResponse> {
  return request({ method: 'PATCH', path: '/portal/v1/preferences', body })
}

File: app/lib/src/api/portal_api.dart
import '../models/portal_models.dart';
import 'agendamento_criado.dart';
import 'portal_client.dart';

/// As chamadas do Portal que o MVP usa, com os tipos que o backend define.
///
/// A tradução de `frontend/src/lib/portal-api.ts`, cortada no escopo do app. Cada
/// caminho e cada parâmetro foi verificado contra o backend em execução — os schemas
/// das rotas são `.strict()`, então **parâmetro desconhecido é 422**, e não algo que o
/// servidor ignora de bom grado.
class PortalApi {
  PortalApi(this._cliente);

  final PortalClient _cliente;

  // ── Antes da sessão ────────────────────────────────────────────────────────

  /// O catálogo de estabelecimentos, para a primeira tela.
  ///
  /// **A única rota do app que não fala de um petshop em particular** — é a pergunta de
  /// quem ainda não escolheu —, e por isso a única fora de `/portal/v1`: todo caminho
  /// daquele prefixo tem o tenant resolvido antes do roteamento, a partir do header do
  /// slug que aqui ainda não existe.
  ///
  /// O que desce é o que está na fachada, e só de quem ligou o Portal do cliente final:
  /// o critério é o mesmo que `tenant()` aplica, então nada daqui abre num 404.
  Future<PortalDirectoryResponse> estabelecimentos() async =>
      PortalDirectoryResponse.fromJson(
        await _cliente.get('/public/v1/portal/tenants', anonimo: true),
      );

  /// A identidade visual do petshop, sem sessão nenhuma.
  ///
  /// É também o que valida o slug que a pessoa digitou: estabelecimento inexistente,
  /// invisível ou em plano sem Portal respondem o mesmo `ERR_PORTAL_001`.
  Future<PortalTenantResponse> tenant() async =>
      PortalTenantResponse.fromJson(await _cliente.get('/portal/v1/tenant', anonimo: true));

  // ── Vínculo da ficha ───────────────────────────────────────────────────────

  /// Pede o código que liga esta conta à ficha que o petshop já tem.
  /// `semNulos` porque `website` é `.optional()` no schema, e não `.nullable()`: o
  /// honeypot da web não existe no app, e mandá-lo como `null` é 422.
  Future<PortalChallengeResponse> pedirCodigo(PortalChallenge entrada) async =>
      PortalChallengeResponse.fromJson(
        await _cliente.post('/portal/v1/access/challenge',
            corpo: semNulos(entrada.toJson())),
      );

  /// Confere o código e conclui o vínculo.
  Future<void> confirmarCodigo(PortalVerify entrada) =>
      _cliente.post('/portal/v1/access/verify', corpo: entrada.toJson());

  // ── A sessão ───────────────────────────────────────────────────────────────

  /// Quem é o tutor aqui, e o que este estabelecimento tem ligado.
  ///
  /// `features` é o que a navegação lê para decidir o que mostrar — agendamento online
  /// e leva-e-traz dependem do plano e da configuração, e o app degrada sozinho em vez
  /// de oferecer uma tela que responderia 402.
  Future<PortalContextResponse> me() async =>
      PortalContextResponse.fromJson(await _cliente.get('/portal/v1/me'));

  // ── Pets ───────────────────────────────────────────────────────────────────

  Future<List<PortalPetSummary>> pets() async {
    final json = await _cliente.get('/portal/v1/pets');
    return (json['pets'] as List).map((p) => PortalPetSummary.fromJson(p)).toList();
  }

  Future<PortalPetDetail> pet(String petId) async =>
      PortalPetDetail.fromJson(await _cliente.get('/portal/v1/pets/$petId'));

  /// O histórico do pet, paginado por cursor.
  Future<PortalTimelineResponse> timeline(String petId, {String? cursor, int? limite}) async =>
      PortalTimelineResponse.fromJson(
        await _cliente.get(
          '/portal/v1/pets/$petId/timeline',
          query: {
            'cursor': ?cursor,
            if (limite != null) 'limit': '$limite',
          },
        ),
      );

  /// A parte da ficha que o tutor pode corrigir: nome, nascimento, castração e
  /// observações. Devolve a ficha **recarregada** — a idade e os alertas não saem de um
  /// `update`, e a tela que acabou de salvar precisa deles.
  ///
  /// **Esta é a chamada em que `semNulos` seria um defeito.** No `UpdateOwnPetSchema`,
  /// `birthDate`, `neutered` e `notes` são `.nullable()` *e* `.optional()`: `null` quer
  /// dizer **apague** e ausente quer dizer **não mexa**. Quem limpou a data de
  /// nascimento pediu para voltar ao "não sei", e cortar o `null` do corpo transformaria
  /// esse pedido em silêncio — o formulário fecharia dizendo que salvou, com o valor
  /// antigo intacto. Peso, porte, raça e pelagem não estão aqui porque não estão no
  /// schema: a trava é o contrato, e o 422 nasce antes do handler.
  Future<PortalPetDetail> atualizarPet(String petId, UpdateOwnPet mudanca) async =>
      PortalPetDetail.fromJson(
        await _cliente.patch('/portal/v1/pets/$petId', corpo: mudanca.toJson()),
      );

  // ── Agendar ────────────────────────────────────────────────────────────────

  /// Os serviços que **este pet** pode receber — a lista depende do porte e da espécie.
  Future<PortalBookingServicesResponse> servicos(String petId) async =>
      PortalBookingServicesResponse.fromJson(
        await _cliente.get('/portal/v1/booking/services', query: {'petId': petId}),
      );

  /// A grade de um dia.
  ///
  /// `dia` é `YYYY-MM-DD` no fuso do petshop, e não um instante: quem escolhe é o dia.
  /// `serviceIds` vai separado por vírgula, como a rota do domínio que responde por
  /// baixo — o BFF repassa a pergunta em vez de recalcular a grade, e é isso que garante
  /// que o tutor veja os mesmos horários que a recepção veria.
  Future<PortalAvailabilityResponse> disponibilidade({
    required String petId,
    required List<String> serviceIds,
    required String dia,
  }) async =>
      PortalAvailabilityResponse.fromJson(
        await _cliente.get('/portal/v1/booking/availability', query: {
          'petId': petId,
          'serviceIds': serviceIds.join(','),
          'date': dia,
        }),
      );

  /// A oferta de leva-e-traz: dá para buscar aqui, e quanto custa cada perna.
  ///
  /// **Não recebe pet**, ainda que a rota viva sob `/booking`: o preço sai do CEP do
  /// endereço primário do **tutor**, e o handler resolve o `tutorId` por
  /// `requireOwnScope`. A primeira versão mandava `?petId=`, que o servidor ignorava em
  /// silêncio — e um parâmetro ignorado é o que faz a tela acreditar numa dependência
  /// que não existe, e repetir a pergunta a cada troca de pet.
  ///
  /// Um POST que criasse a corrida para descobrir o preço deixaria lixo no painel do
  /// petshop a cada pergunta do tutor; é para isso que esta rota existe.
  Future<PortalTaxiOffer> ofertaDeTaxi() async =>
      PortalTaxiOffer.fromJson(await _cliente.get('/portal/v1/booking/taxi'));

  /// Marca o horário.
  ///
  /// `semNulos` porque `notes` e `taxi` são `.optional()`: um pedido sem observação nem
  /// leva-e-traz não deve dizer `null` — deve não dizer nada.
  ///
  /// **A resposta não é `PortalAppointmentDetail`**, ainda que se pareça: é
  /// `AgendamentoCriado`, a única classe escrita à mão do app, porque esta rota não tem
  /// schema Zod. A confusão durou da fatia 1 até `tool/smoke_booking.dart` marcar um
  /// horário de verdade e o `actions` ausente derrubar o `fromJson`.
  Future<AgendamentoCriado> agendar(PortalBooking pedido) async =>
      AgendamentoCriado.fromJson(
        await _cliente.post('/portal/v1/booking', corpo: semNulos(pedido.toJson())),
      );

  // ── Agendamentos ───────────────────────────────────────────────────────────

  Future<PortalAppointmentsResponse> agendamentos({String? cursor, int? limite}) async =>
      PortalAppointmentsResponse.fromJson(
        await _cliente.get('/portal/v1/appointments', query: {
          'cursor': ?cursor,
          if (limite != null) 'limit': '$limite',
        }),
      );

  Future<PortalAppointmentDetail> agendamento(String id) async =>
      PortalAppointmentDetail.fromJson(await _cliente.get('/portal/v1/appointments/$id'));

  /// Cancela, e devolve o agendamento já cancelado.
  ///
  /// Fora da janela, a **primeira** tentativa é recusada com `ERR_PORTAL_011` dizendo
  /// quanto custa, e só a segunda — com `acknowledgeFee` — passa. A consequência é
  /// mostrada antes, e quem confirma é a pessoa, não a tela. A tela pode estar velha; a
  /// regra não, e é por isso que o servidor recusa mesmo quando o botão já avisou.
  Future<PortalAppointmentDetail> cancelar(String id, {bool aceitarTaxa = false}) async =>
      PortalAppointmentDetail.fromJson(
        await _cliente.post(
          '/portal/v1/appointments/$id/cancel',
          corpo: PortalCancel(acknowledgeFee: aceitarTaxa).toJson(),
        ),
      );

  /// Remarca. Devolve o agendamento **novo** — o anterior vira histórico no instante
  /// da resposta, e é para o novo que a tela olha.
  Future<PortalAppointmentDetail> remarcar(String id, PortalReschedule destino) async =>
      PortalAppointmentDetail.fromJson(
        await _cliente.post('/portal/v1/appointments/$id/reschedule', corpo: destino.toJson()),
      );

  // ── Financeiro ─────────────────────────────────────────────────────────────

  /// O painel da conta: saldo, pacotes com crédito e como pagar (MOD-PORTAL-08).
  ///
  /// `balanceCents` é **negativo para dívida** — a convenção da plataforma, e a mesma de
  /// `tutors.balance_cents`. Nenhuma tela refaz essa leitura à mão: `deveEmCentavos` e
  /// `creditoEmCentavos`, em `dinheiro.dart`, são a tradução das duas funções que
  /// `shared-types/portal.ts` criou depois de o Portal da web ter dito "Sem pendências"
  /// a quem devia.
  Future<PortalFinanceResponse> financeiro() async =>
      PortalFinanceResponse.fromJson(await _cliente.get('/portal/v1/finance'));

  /// O extrato, paginado **por página** e não por cursor.
  ///
  /// É a única lista do app assim, e a razão está no dado: o extrato ordena por
  /// `occurred_at`, que repete — três serviços do mesmo dia entram no mesmo instante —,
  /// e um cursor por data pularia ou repetiria linhas. A resposta traz o `total`, que é
  /// como a tela sabe quando parar de oferecer "Ver mais".
  Future<PortalStatementResponse> extrato({int? pagina, int? limite}) async =>
      PortalStatementResponse.fromJson(
        await _cliente.get('/portal/v1/finance/statement', query: {
          if (pagina != null) 'page': '$pagina',
          if (limite != null) 'limit': '$limite',
        }),
      );

  /// O mesmo extrato em papel (AC-02 de MOD-DOC-09).
  ///
  /// Bytes, e não uma URL: o extrato não é arquivado, então não há endereço a assinar.
  /// O nome de reserva tem a data para o caso de o `content-disposition` não chegar —
  /// dois extratos baixados no mesmo aparelho precisam de nomes diferentes.
  Future<ArquivoDoPortal> extratoEmPdf(DateTime hoje) => _cliente.arquivo(
        '/portal/v1/finance/statement/pdf',
        nomePadrao: 'extrato-${hoje.toIso8601String().substring(0, 10)}.pdf',
      );

  /// O recibo de um pagamento (AC-03 de MOD-PORTAL-08).
  ///
  /// `url` **pode voltar nula**, e isso não é erro: o PDF nasce depois do pagamento,
  /// fora da transação, e um recibo ainda em preparo tem número e não tem arquivo. A
  /// tela diz "em preparo" em vez de abrir um endereço morto.
  ///
  /// 404 aqui é pagamento que não é deste tutor (RN-03), e a tela **não** o distingue de
  /// uma falha: dizer "não é seu" a quem adivinhou um id confirmaria que ele existe.
  Future<PortalReceiptResponse> recibo(String paymentId) async =>
      PortalReceiptResponse.fromJson(
        await _cliente.get('/portal/v1/finance/receipts/$paymentId'),
      );

  // ── Meus Dados ─────────────────────────────────────────────────────────────
  //
  // **Toda escrita desta seção devolve a ficha inteira relida**, e não o recurso que
  // mudou: a tela é uma só, e é a resposta do servidor que passa a valer — a tela nunca
  // remenda o que tinha com o que supôs ter salvado.

  /// A ficha como o titular a vê (MOD-PORTAL-09). Sem `notes`: aquilo é o caderno da
  /// recepção, e só desce na exportação.
  Future<PortalMeDataResponse> meusDados() async =>
      PortalMeDataResponse.fromJson(await _cliente.get('/portal/v1/me/data'));

  /// Nome social e nascimento — os dois campos que o tutor muda sozinho.
  ///
  /// **Sem `semNulos`, pela mesma razão do `atualizarPet`:** no `UpdateOwnTutorSchema`
  /// os dois são `.nullable()`, e `null` é o "apague" de quem limpou o campo. O
  /// formulário manda os dois sempre. Nome civil, CPF e contato ficam fora porque o
  /// schema é `.strict()` — mandá-los seria 422.
  Future<PortalMeDataResponse> atualizarPerfil(UpdateOwnTutor mudanca) async =>
      PortalMeDataResponse.fromJson(
        await _cliente.patch('/portal/v1/me/data', corpo: mudanca.toJson()),
      );

  /// Endereço novo. `semNulos` porque `complement` e `accessNotes` são `.optional()`, e
  /// não `.nullable()`: aqui `null` é 422.
  Future<PortalMeDataResponse> adicionarEndereco(PortalAddressInput endereco) async =>
      PortalMeDataResponse.fromJson(
        await _cliente.post('/portal/v1/me/addresses', corpo: semNulos(endereco.toJson())),
      );

  /// Corrige um endereço.
  ///
  /// Também `semNulos`, e o jeito de **apagar** o complemento é string vazia, não
  /// `null`: o serviço de endereços grava `null` quando recebe `""`. Quem monta o
  /// pedido é o formulário, que sabe a diferença entre "não mexi" e "limpei".
  Future<PortalMeDataResponse> corrigirEndereco(
    String enderecoId,
    UpdatePortalAddress mudanca,
  ) async =>
      PortalMeDataResponse.fromJson(
        await _cliente.patch(
          '/portal/v1/me/addresses/$enderecoId',
          corpo: semNulos(mudanca.toJson()),
        ),
      );

  /// Pede o código que confirma um telefone ou e-mail novo (AC-02).
  ///
  /// O código sai para o contato **novo** — é a posse dele que se prova. Pedir de novo
  /// invalida o pedido anterior no servidor, e é por isso que a tela deixa voltar e
  /// corrigir o número digitado mesmo com um desafio aberto.
  Future<PortalContactChangeResponse> pedirTrocaDeContato(PortalContactChange troca) async =>
      PortalContactChangeResponse.fromJson(
        await _cliente.post('/portal/v1/me/contact', corpo: troca.toJson()),
      );

  /// Confere o código. Só agora o contato entra na ficha.
  Future<PortalMeDataResponse> confirmarTrocaDeContato(PortalContactVerify codigo) async =>
      PortalMeDataResponse.fromJson(
        await _cliente.post('/portal/v1/me/contact/verify', corpo: codigo.toJson()),
      );

  /// A cópia dos dados em PDF (AC-04 — LGPD art. 18, direito de acesso).
  ///
  /// Bytes, como o extrato: o documento é o retrato da ficha agora, e guardá-lo num
  /// bucket criaria uma segunda cópia dos dados pessoais só para poder entregá-los.
  /// Cada chamada vira `tutor.exported` na trilha — é a prova de que o direito foi
  /// exercido.
  Future<ArquivoDoPortal> meusDadosEmPdf(DateTime hoje) => _cliente.arquivo(
        '/portal/v1/me/export/pdf',
        nomePadrao: 'meus-dados-${_diaDoArquivo(hoje)}.pdf',
      );

  /// A mesma exportação em JSON — a portabilidade do art. 19, o arquivo que outro
  /// sistema consegue importar.
  ///
  /// Vai por `arquivo`, e não por `get`: o app não lê este JSON, entrega-o. Decodificar
  /// e codificar de novo só arriscaria mudar o que o servidor escreveu. A rota não manda
  /// `content-disposition`, então o nome é sempre o de reserva.
  Future<ArquivoDoPortal> meusDadosEmJson(DateTime hoje) => _cliente.arquivo(
        '/portal/v1/me/export',
        nomePadrao: 'meus-dados-${_diaDoArquivo(hoje)}.json',
      );

  /// O pedido de exclusão (AC-05). **Registra, não apaga**: vira uma linha na fila da
  /// equipe, que responde em até 15 dias. `semNulos` porque o motivo é opcional, e
  /// ausente — não `null`.
  Future<PortalMeDataResponse> pedirExclusao(PortalDeletionRequestInput pedido) async =>
      PortalMeDataResponse.fromJson(
        await _cliente.post('/portal/v1/me/deletion-request', corpo: semNulos(pedido.toJson())),
      );

  static String _diaDoArquivo(DateTime dia) => dia.toIso8601String().substring(0, 10);
}
